### DISCUSSION TITLE
What is the reason to call convertToDatabaseValue() during hydration?

### DISCUSSION THREAD

**Discussion by lagnat** [mikro-orm/mikro-orm#3754] [Q&A] [answered]
Here's a snippet of a generated hydrator.

```
    if (convertCustomTypes) {
      const value = convertToJSValue_name(data.name);
      data.name = convertToDatabaseValue_name(value);  <--- this
      entity.name = value;
    } else {
      entity.name = data.name;
    }
```

I totally didn't expect convertToDatabaseValue() to be called during a fetch from the database.  Can you clarify why this is here?

**Comment by B4nan** [comment, marked as answer]
Because we need the database value in `data`, and we don't have it all the time - some code paths (e.g. with joined loading strategy) end up with different `data` shape, and this normalizes it.

**Reply by lagnat** [reply]
I don't understand how the internals work but I'm satisfied that this is intentional and necessary.  Thank you!

**Reply by B4nan** [reply]
Intentional, but maybe we could get around the need for it.

**Reply by lagnat** [reply]
For context, since I'm doing input validation in custom types, some bad data in the DB is throwing errors when I do the query.  One could argue that bad data should not be allowed to reach the database or it should be migrated out but that's often not as simple as it sounds.  Prefer to have the query succeed.  I can choose to do something about bad data in `convertToJSValue()` if I choose.
